Merge my initial code into develop

AaroSnid opened the pull request.

Commit:
Added Google Calendar Functionality

Commit:
Added Google Calendar Functionality

AaroSnid committed:
Rebase the source files

AaroSnid committed:
Rebase the source files

AaroSnid committed:
Merge branch 'dealWithThisPlease' of github.com:AaroSnid/Bibb into dealWithThisPlease

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+import win32gui
+import time
+
+'''This is to designate different "tasks" or study topics. Defininately adding more variables and methods.'''
```
*definately

RRickyH commented on credentials.json:
```diff
@@ -0,0 +1 @@
+{"installed":{"client_id":"621358017264-rrecp379ff662ikk4bi6613nuko91i1t.apps.googleusercontent.com","project_id":"bibb-456220","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"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","redirect_uris":["http://localhost"]} ...
```
Pretty printed please? 🥺

AaroSnid replied:
Given the user is, in an ideal world, never going to see this, I don't see why we would make this easier to access.

RRickyH commented on src_files/calendars.py:
```diff
@@ -0,0 +1,64 @@
+import datetime
```
Imported twice?

RRickyH replied:
Not necessary.

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, name, action_type, whitelist):
+        self.name = name  # e.g. “Calculus”
+        self.action_type = action_type  # e.g. “Studying”
+        self.whitelist = whitelist  # Set of permitted applications
```
What is "whitelist"? Maybe add a whitelist class with add and remove methods to match usage further down.

AaroSnid replied:
Maybe read the comment and it will tell you

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+def event_loop(Event, Calendar):
+    initial_time = time.time()
+    event_details = {'start': {'dateTime': datetime.now(timezone.utc).replace(microsecond=0).isoformat()},
+                     'summary': Event.action_type + ' ' + Event.name}
```
Maybe format this dictionary like:
'''python
event_details = {
    "start": {
        "datetime": datetime.now(timezone.utc).replace(microsecond=0).isoformat()
        },
    "summary": f"{Event.action_type} {Event.name}"
}
'''

RRickyH replied:
datetime is not imported :(

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+'''This will be ran whenever a session is begun'''
+
+
+def event_loop(Event, Calendar):
```
You can't use Event and Calendar as parameters since these names shadow classes defined above.

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+'''This will be ran whenever a session is begun'''
+
+
+def event_loop(Event, Calendar):
```
I think event loop will be handled by the PySide6 framework.

AaroSnid replied:
Sounds good to me. I've kept my function in there for reference when that is written, but I'm fine letting it be removed.

RRickyH commented on src_files/calendars.py:
```diff
@@ -0,0 +1,64 @@
+class GoogleCalendar(Calendar):
+
+    def get_credentials(self):
+        #The base code for this function comes from the Python Quickstart at https://developers.google.com/workspace/calendar/api/quickstart/python, published under Apatche 2.0 license
```
This line is way too long.

RRickyH commented on src_files/calendars.py:
```diff
@@ -0,0 +1,64 @@
+        event = {
+            'summary': event_details['summary'],
+            'description': event_details['description'],
+        'start': {
```
missing indent.

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+import calendars.py
```
You don't put .py extensions in imports.

AaroSnid replied:
Thanks for being patient

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+
+    # When the main event loop ends
+    else:
+        event_details.update({'end': {'dateTime': datetime.now(timezone.utc).replace(microsecond=0).isoformat()}})
```
timezone is also not imported.

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+
+        # Checks to make sure the user is on topic
+        window = get_window_name()
+        if window not in Event.whitelist:
```
Note to define a __contains__ method for our whitelist class?

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+import calendars.py
+import time_tracking.py
+
+import win32gui
```
This is unused.

RRickyH commented on src_files/events.py:
```diff
@@ -0,0 +1,68 @@
+        if desc:
+            event_details.update({'description': desc})
+
+        Calendar.create_event()
```
Calendar is a class here and doesn't have a class method named create_event... it requires a 'self'.

RRickyH commented on src_files/time_tracking.py:
```diff
@@ -0,0 +1,23 @@
+    if user_input == 1:
+        Event.update_whitelist(window, "add")
+    elif user_input == 2:
+        win32gui.PostMessage(window, win32con.WM_CLOSE, 0, 0)
```
I think win32con was supposed to by win32gui.

RRickyH commented on src_files/time_tracking.py:
```diff
@@ -0,0 +1,23 @@
+    return win32gui.GetWindowText(window)
+
+#Called when user is not on topic, and used to select action
+def on_topic_reminder(Event):
```
Same thing here with this parameter. It should be "event".

RRickyH reviewed (changes requested):
I feel like we should keep out GUI, focus mode, and time tracking for now.

AaroSnid committed:
Resolving pre-merge issues

RRickyH commented on src_files/events.py:
```diff
@@ -1,56 +1,69 @@
 
-    # This will probably never be ran
+        # Args to be used if class is made from database information.
+        if args:
```
It might make more sense to use kwargs here. I feel like that would prevent people from accidentally putting these in the wrong order and also make it a bit more readable. It would also cleanup the if statement since you could just do something like `self.goal_period = kwargs.get("goal_period")`

AaroSnid replied:
I think were thinking of these in different terms. In my mind, the user will never use any of these functions, and it will only be the program that enters the information, so we only have to worry about us getting the calls right. I do however agree that kwargs would make more sense here. Just saying this for an explanation of my thought process.

RRickyH replied:
I agree that the end user would never call this function but I was thinking about us. Developers probably use the code much more than users do and it's important to make things as clear as possible. It's best if you can get things to a state where you would not need to look inside of the function or know how it works in order to use it.

RRickyH commented on src_files/events.py:
```diff
@@ -60,9 +73,9 @@ def event_loop(Event, Calendar):
     else:
         event_details.update({'end': {'dateTime': datetime.now(timezone.utc).replace(microsecond=0).isoformat()}})
```
This needs another indent.

AaroSnid replied:
What exactly here does? The else statement is supposed to be executed when the while loop ends, unless you are talking about something else.

RRickyH replied:
I was not aware that python while loops had else statements and I thought this was supposed to be off of the if statement. The else statement is redundant though since there are no break statements in the loop. Here's what pylint gives me:
```commandline
src_files/events.py:73:4: W0120: Else clause on loop without a break statement, remove the else and de-indent all the code inside it (useless-else-on-loop)
```

RRickyH commented on src_files/events.py:
```diff
@@ -1,56 +1,69 @@
+            time_tracking.on_topic_reminder(Event)
             paused_time += (time.time() - temp_time)
 
         # Calculates elapsed time, can be displayed to user
```
Time elapsed isn't being used for anything.

AaroSnid replied:
I know. It's something we definitely need to measure but we don't have a use programmed for it yet. I can comment it out if you want.

RRickyH replied:
That's alright, I just wanted to make sure it was intentional since pylint was complaining about this.

RRickyH reviewed (commented):
Hey, thanks for making these changes! I did miss a couple things my last read through and I'm also running into a circular import error. I think we could get rid of this with
```python
from typing import TYPE_CHECKING
more imports ...

if TYPE_CHECKING:
    problematic imports ...
```
[Why if TYPE_CHECKING](https://vickiboykis.com/2023/12/11/why-if-type_checking/)

I feel like we could also get rid of event loop and stash it somewhere else for reference since it doesn't have much to do with calendars anyhow.

RRickyH commented on src_files/events.py:
```diff
@@ -1,56 +1,69 @@
+        if window not in event.whitelist:
             temp_time = time.time()
-            on_topic_reminder(Event)
+            time_tracking.on_topic_reminder(Event)
```
I think you forgot to change this one to be an event instance.

AaroSnid committed:
Bug fixes and preparing for merge

The circular import error should be fixed, as only one of the used files
import the others. If not, let me know.

RRickyH commented on src_files/calendars.py:
```diff
@@ -0,0 +1,89 @@
+        events = events_result.get("items", [])
+        return events
+
+    def create_event(self, event_details:dict):
```
I'm not sure why we aren't using the event class here. I feel like dictionaries are a little too unwieldy for storing event details, there's too much freedom.  Looking up the documentation is also a real pain.

AaroSnid replied:
Ill make a note to change this for the next push.

RRickyH commented on src_files/calendars.py:
```diff
@@ -0,0 +1,89 @@
+    def create_event(self, event_details:dict):
+        pass
+
+class GoogleCalendar(Calendar):
```
+1 Tested

RRickyH reviewed (approved):
This is a great start for our project. I tested the GoogleCalendar object and its methods are working as expected.

AaroSnid commented:
I cannot state how happy I am to have that horrible branch name gone
